graphql2/graphqlapp/calendarsubscription.go

```diff
@@ -24,8 +25,17 @@ func (a *UserCalendarSubscription) Schedule(ctx context.Context, obj *calendarsu
 	return a.ScheduleStore.FindOne(ctx, obj.ScheduleID)
 }
 func (a *UserCalendarSubscription) URL(ctx context.Context, obj *calendarsubscription.CalendarSubscription) (*string, error) {
-	// URL generation out-of-scope at this stage
-	return nil, validation.NewFieldError("URL", "not implemented")
+	tok := obj.Token()
+	if tok == "" {
+		return nil, nil
+	}
+
+	v := make(url.Values)
+	v.Set("token", tok)
+
+	cfg := config.FromContext(ctx)
+	callback := cfg.CallbackURL("/api/v2/calendar", v)
```

Comment:
This is intentional, I'll elaborate as I think it's been more of an unwritten thing and I think the context could be helpful.

---

Now, we can't import from `inithttp.go` because that would cause an import cycle. But if we move it to cal-sub then `inithttp.go` won't contain the route string. You'd have to look at the const every time you're looking at routes. There's a value-trade here.

The indirection of a const isn't particularly helpful in this case, because we're prioritizing readability above change-ability. Our URL paths need to stay permanent once released, and in the case, that we absolutely need to change this URL (e.g. like the v1 -> v2 migration), we would need to change the two one-at-a-time: by adding a new route in `inithttp.go` in one version, then changing callback generation in a future version, and requiring admins to update to version X in full before version Y.

`inithttp.go` should always contain easily readable/scannable text of all routes so it's quick/easy to find handlers in one place. All generated callbacks go through `CallbackURL` so it's easy to find all generated URLs passed to clients by searching for the usage of that method. We could name the const `APIV2Calendar` or something that looks like the path, but that's less readable than just seeing `"/api/v2/calendar"` IMO.

The value of ensuring they always match, however, is handled by smoketests (to be added in feature branch once the iCal generation is merged). Whatever URL we get back from the mutation, should give iCal data. Additionally, a test should include it's own hard-coded `/api/v2/calendar` string because it's a failure if that route changes and breaks, even if a new one is added. Once the code is written and working the first time, the primary value is in debugging and searching.

We get the same guarantees of correctness, without the indirection, the code is more searchable (e.g. when correlating to logs/paths), and readable because you always see what's happening directly. Searching for `/api/v2/calendar` requires no guessing at const names, it's the string directly from the logs, and you can quickly find every single possible entry or generation point by doing a text search.

---

That's why we've kept API paths duplicated like this for all callbacks, rather than putting them all in const values. It's one of those times where a little copying is better than a little dependency.

Reply:
Thanks for the write-up, Nate! This helped me understand as well.